Taking stop-signal into account when docker kill

vdemeester opened the pull request:
**\- What I did**

This fix tries to address the issue #11065, taking into account `stop-signal` if present so that non-fatal signal (i.e. `SIGKILL`) is sent using `docker kill` command, it will not disable the restart policy. 🐯

**\- How to verify it**

Added 2 integration tests to handle those cases.

**\- Still todo**
- [x] Validate `SIGKILL` behavior
- [ ] Update documentation

This fix fixes #11065.

/cc @tonistiigi @crosbymichael 

🐸

Signed-off-by: Vincent Demeester vincent@sbr.pm

vdemeester commented:
/cc @mlaventure @thaJeztah @cpuguy83 :angel:

mlaventure commented on integration-cli/docker_cli_kill_test.go:
```diff
@@ -47,7 +47,7 @@ func (s *DockerSuite) TestKillDifferentUserContainer(c *check.C) {
 // regression test about correct signal parsing see #13665
 func (s *DockerSuite) TestKillWithSignal(c *check.C) {
-	// Cannot port to Windows - does not support signals in the same was a Linux does
+	// Cannot port to Windows - does not support signals in the same way a Linux does
```
nit: "same way ~~a~~ Linux does"

vdemeester replied:
:sweat_smile:

mlaventure commented on daemon/kill.go:
```diff
@@ -105,6 +115,9 @@ func (daemon *Daemon) Kill(container *container.Container) error {
 		return errNotRunning{container.ID}
 	}
 
+	// FIXME(vdemeester) should we disable restartpolicy for SIGKILL or not ?
```
From my POV I would expect the RestartPolicy to only be disabled if the kill was generated as part of a `docker stop` command.

WDYT?

vdemeester replied:
@mlaventure you mean the SIGKILL sent after the grace period right ?


I'm not really sure in fact. Naively, I would consider SIGKILL as the *fatal* signal, and thus, the only signal that always disable restart policy.

mlaventure replied:
But the `SIGKILL` could have been sent as part of the `oom` policy in which case, I'd probably want my container restarted since the targeted process are more or less random

vdemeester replied:
@mlaventure hum… true, so `SIGKILL` would be treated exactly as the other signals then. `docker stop` sends the `stop-signal` if present, and thus, the restart manager would disable the restart policy, even if the `SIGKILL` is sent afterwards.

@crosbymichael @LK4D4 @cpuguy83 @thaJeztah WDYT ? I can update to reflect that.

thaJeztah replied:
I need to read up on this issue again :smile:, and see what's the most logical behavior here

thaJeztah replied:
So, this is only in the case of a _shutdown_, correct? I think a shutdown should not influence the restart policy, independent of the signal that was used (i.e. if the container was running, and it's killed as part of the shutdown, the _intent_ was to have it running, so `--restart=unless-stopped` still applies).

thaJeztah replied:
Ok after some discussion; it's better to keep the current behavior here; can your remove this?

vdemeester replied:
I can :stuck_out_tongue_closed_eyes:

vdemeester committed:
Taking stop-signal into account when docker kill

If a container sets a stop-signal, taking it into account to disable or
not the restart policy.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester commented:
PR updated, withouth changes on `kill` :stuck_out_tongue:

mlaventure commented on daemon/kill.go:
```diff
@@ -68,7 +68,17 @@ func (daemon *Daemon) killWithSignal(container *container.Container, sig int) er
+			container.ExitOnNext()
+		}
+	} else {
+		container.ExitOnNext()
```
Just realized, shouldn't we check that the `signal` being sent is the `stop` signal here too before disabling the restartmanager?

vdemeester replied:
If the container has no `StopSignal` then doing the same as before (for retro-compatibility sake :sweat_smile:) :angel:

mlaventure replied:
Yes, I noticed it was the same before. It just feels weird. This can be done in a subsequent PR anyhow I guess.

cc @tonistiigi

crosbymichael commented:
LGTM
